src/System.Net.HttpListener/tests/HttpListenerContextTests.cs

```diff
@@ -0,0 +1,178 @@
+        [InlineData("random}text")]
+        [InlineData("\x19")]
+        [InlineData("\x7f")]
+        public async Task AcceptWebSocketAsync_InvalidSubProtocol_ThrowsArgumentException(string subProtocol)
+        {
+            HttpListenerContext context = await GetWebSocketContext();
+            await Assert.ThrowsAsync<ArgumentException>("subProtocol", () => context.AcceptWebSocketAsync(subProtocol));
+        }
+
+        [ConditionalTheory(nameof(IsNotWindows7OrUapCore))]
+        [InlineData("!")]
+        [InlineData("#")]
+        [InlineData("YouDontKnowMe")]
+        public async Task AcceptWebSocketAsync_NoSuchSubProtocol_ThrowsWebSocketException(string subProtocol)
+        {
+            HttpListenerContext context = await GetWebSocketContext();
+            await Assert.ThrowsAsync<WebSocketException>(() => context.AcceptWebSocketAsync(subProtocol));
+        }
+
+        [ConditionalFact(nameof(IsNotWindows7OrUapCore))]
+        public async Task AcceptWebSocketAsync_InvalidKeepAlive_ThrowsWebSocketException()
+        {
+            HttpListenerContext context = await GetWebSocketContext();
+
+            TimeSpan keepAlive = TimeSpan.FromMilliseconds(-2);
+            await Assert.ThrowsAsync<ArgumentOutOfRangeException>("keepAliveInterval", () => context.AcceptWebSocketAsync(null, keepAlive));
+        }
+
+        [ConditionalTheory(nameof(IsNotWindows7OrUapCore))]
+        [InlineData(-1)]
+        [InlineData(0)]
+        [InlineData(255)]
+        [InlineData(64 * 1024 + 1)]
+        public async Task AcceptWebSocketAsync_InvalidReceiveBufferSize_ThrowsWebSocketException(int receiveBufferSize)
+        {
+            HttpListenerContext context = await GetWebSocketContext();
+            await Assert.ThrowsAsync<ArgumentOutOfRangeException>("receiveBufferSize", () => context.AcceptWebSocketAsync(null, receiveBufferSize, TimeSpan.MaxValue));
+        }
+
+        private async Task GetSocketContext(string[] headers, Action<HttpListenerContext> contextAction)
+        {
+            using (Socket client = Factory.GetConnectedSocket())
+            {
+                client.Send(Factory.GetContent("1.1", "POST", null, "Text", headers, true));
+
+                HttpListener listener = Factory.GetListener();
+                contextAction(await listener.GetContextAsync());
+            }
+        }
+
+        private Task<HttpListenerContext> GetWebSocketContext(string[] subProtocols = null)
+        {
+            if (subProtocols != null)
+            {
+                foreach (string subProtocol in subProtocols)
+                {
+                    Socket.Options.AddSubProtocol(subProtocol);
+                }
+            }
+
+            var uriBuilder = new UriBuilder(Factory.ListeningUrl) { Scheme = "ws" };
+            Task<HttpListenerContext> serverContextTask = Factory.GetListener().GetContextAsync();
+
+            Task clientConnectTask = Socket.ConnectAsync(uriBuilder.Uri, CancellationToken.None);
```

Comment:
This this connection fails, the test will end up hanging, even though we could have accurately reported the failure that was stored in the clientConnectTask.  That's what was happening in the test I fixed here: https://github.com/dotnet/corefx/pull/19755